### PR TITLE
Update thrift to v0.21.0

### PR DESCRIPTION
**Pull request by fishy**
Also rerun thrift compiler v0.21.0 on baseplate.thrift.

Also add thriftbp.ClientPoolConfig.UseZlib, which was not possible before.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1,11 +1,11 @@
 module github.com/reddit/baseplate.go
 
-go 1.22
+go 1.22.0
```
Comment (mathyourlife-reddit) [inline go.mod:3]: this change necessary?

**Review by mathyourlife-reddit — APPROVED** [review]
LGTM. [Go changes](https://github.com/apache/thrift/blob/master/CHANGES.md#0210) seem fairly limited.

File: go.mod (hunk 1 above)
Comment (fishy) [reply go.mod:3]: I believe it is, as thrift declared `1.22.0` in v0.21.0 https://github.com/apache/thrift/blob/v0.21.0/go.mod#L3, and only `1.22.0` is considered matching that, and `1.22` does not.

it won't be required once we update to 1.23 here, and in the next version of thrift I'll also remember to remove the patch version.

File: thriftbp/client_pool.go [hunk 2]
```diff
@@ -228,6 +228,12 @@ type ClientPoolConfig struct {
 	//
 	// Optional. If empty, no "thrift-hostname" header will be sent.
 	ThriftHostnameHeader string `yaml:"thriftHostnameHeader"`
+
+	// Uses zlib transform in thrift THeader connections between client and
```
Comment (kylelemons) [inline thriftbp/client_pool.go:232]: is THeader just for headers, or for the whole connection?  `UseZlib` sounds like this is for the whole connection, but the docs make it sound like it's just for headers...

so maybe
* rename to `UseZlibForHeaders`, or
* update the comment to specify that this is for the whole connection

?

**Review by kylelemons — APPROVED** [review]

File: thriftbp/client_pool.go (hunk 2 above)
Comment (fishy) [reply thriftbp/client_pool.go:232]: `THeader` is a thrift TProtocol/TTransport (it's just the only TProtocol/TTransport that supports headers, thus the name)

so it's for the whole connection.

**Review by konradreiche — APPROVED** [review]

**Review by konradreiche — APPROVED** [review]

**Pull request merged**